src/engine/engine.cpp

```diff
@@ -99,11 +103,32 @@ RunQuery(const std::unique_ptr<osrm::engine::Engine::EngineLock> &lock,
     BOOST_ASSERT(lock);
     lock->IncreaseQueryCount();
 
-    auto &shared_facade = static_cast<osrm::engine::datafacade::SharedDataFacade &>(*facade);
-    shared_facade.CheckAndReloadFacade();
-    // Get a shared data lock so that other threads won't update
-    // things while the query is running
-    boost::shared_lock<boost::shared_mutex> data_lock{shared_facade.data_mutex};
+    {
+        // this lock ensures that we are never overtaken while creating a new
+        // facade and setting it.
+        // This is important since we need to ensure there is always exactly
+        // one facade per shared memory region.
+        // TODO: Remove this once the SharedDataFacade doesn't own the shared memory
+        // segment anymore.
+        std::lock_guard<std::mutex> update_lock(lock->facade_update_mutex);
```

Comment:
Ah! This is important, I already was wondering about the `shared_ptr &` above! Thanks for the comment!